### PR TITLE
Adjust SVG Viewbox to contain whole letter

### PR DESCRIPTION
**Pull request by davidgodzsak**
SVG ViewBox originally contained glyphs from the baseline, with a `-av` flag this can now be changed to also include the descender and lineGap.

### CODE REVIEW

File: lib/opentypesvg/fonts2svg.py [hunk 1]
```diff
@@ -126,20 +126,15 @@ def processFonts(font_paths_list, hex_colors_list, outputFolderPath, options):
...
-    # the same UPM value. If fetching the UPM value fails, default to 1000.
-    try:
-        upm = ttLib.TTFont(font_paths_list[0])['head'].unitsPerEm
-    except KeyError:
-        upm = 1000
-
     nestedFolderPath = None
     filesSaved = 0
 
+    viewbox = viewbox_settings(font_paths_list, options)
```
Comment (miguelsousa) [inline lib/opentypesvg/fonts2svg.py:132]: Pass only the first font, and only the option that matters.

File: lib/opentypesvg/fonts2svg.py [hunk 2]
```diff
@@ -193,6 +188,31 @@ def processFonts(font_paths_list, hex_colors_list, outputFolderPath, options):
...
+            upm = tt_font['head'].unitsPerEm
+        except KeyError:
+            upm = 1000
+        return """viewBox="0 -{} {} {}">\n""".format(upm, upm, upm)
+    else:
+        try:
+            hhea = tt_font['hhea']
+            height = hhea.ascender + abs(hhea.descender) + hhea.lineGap # hhea.descender is negative
+            width = hhea.advanceWidthMax
+            y = -hhea.ascender
```
Comment (miguelsousa) [inline lib/opentypesvg/fonts2svg.py:206]: Use the x/yMin/Max values from the `head` table instead.

Comment (miguelsousa) [reply lib/opentypesvg/fonts2svg.py:206]: Also, only one try/except is needed; move the if/else branching to the inside of it.

File: lib/opentypesvg/fonts2svg.py [hunk 3]
```diff
@@ -260,6 +280,12 @@ def get_options(args):
         dest='glyphsets_union',
         help="do union (instead of intersection) of the fonts' glyph sets."
     )
+    parser.add_argument(
+        '-av', '--adjust-viewbox',
+        action='store_true',
+        dest='adjust_view_box_to_glyph',
+        help="adjust the viewbox to the glyph (descender calculated and lineGap in height)."
```
Comment (miguelsousa) [inline lib/opentypesvg/fonts2svg.py:287]: This description is confusing. What about "vertically center the viewBox on the bounding box of all glyphs" ?

**Review by miguelsousa — CHANGES_REQUESTED** [review]
In addition to the inline comments, there are several code formatting problems. See the CI errors, or use flake8 locally.

**Comment by davidgodzsak** [comment]
After some manual testing, I noticed that `head.yMin` possibly denotes the bottom of the bounding box for a glyph and `yMax` the top while in SVG it is the other way around, hence `y = -head.yMax`.

File: tests/fonts2svg_test.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+from opentypesvg.fonts2svg import viewbox_settings
+
+
+def test_adjust_to_viewbox(base_font_path, capsys):
+    viewbox = viewbox_settings(base_font_path, True)
+    assert viewbox == "-168 -850 998 1151"
+
+
+def test_adjust_to_viewbox_default(base_font_path, capsys):
```
Comment (miguelsousa) [inline tests/fonts2svg_test.py:11]: You can remove `capsys` on both tests.

**Review by miguelsousa — CHANGES_REQUESTED** [review]
> I noticed that `head.yMin` possibly denotes the bottom of the bounding box for a glyph and `yMax` the top while in SVG it is the other way around

Right, [in SVG the coordinate system is `y-down`](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Positions#the_grid) whereas the font's is the common `y-up`.

> hence `y = -head.yMax`

I think you'll have to factor `head.yMin` somewhere because if the font has a `head.yMin` value grater than zero the glyphs won't be centered nor scaled as expected. I made these two edge-case fonts to help you test.

[font_registerd.ttf.zip](https://github.com/adobe-type-tools/opentype-svg/files/6536718/font_registerd.ttf.zip)
[font_trademark.ttf.zip](https://github.com/adobe-type-tools/opentype-svg/files/6536720/font_trademark.ttf.zip)

**Comment by davidgodzsak** [comment]
> I think you'll have to factor `head.yMin` somewhere because if the font has a `head.yMin` value grater than zero the glyphs won't be centered nor scaled as expected. I made these two edge-case fonts to help you test.

I already have them factored in the:  `height = head.yMax + abs(head.yMin)`.  I think this is an solution for the y-up -y-down problem together with having -yMax as the top

**Review by miguelsousa — CHANGES_REQUESTED** [review]
As far as I can tell, you've got the values of `x` and `y` nailed, but `width` and `height` need some tweaks.

With the current code, the SVG that will come out of **[font_registerd.ttf](https://github.com/adobe-type-tools/opentype-svg/files/6536718/font_registerd.ttf.zip)** is this,
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="23 -713 423 1032">
	<path fill="#000000" d="M211 -319Q160 -319 117 -344Q74 -369 48.5 -413.5Q23 -458 23 -516Q23 -575 48.5 -619Q74 -663 117 -688Q160 -713 211 -713Q263 -713 306 -688Q349 -663 374.5 -619Q400 -575 400 -516Q400 -458 374.5 -413.5Q349 -369 306 -344Q263 -319 211 -319ZM211 -356Q253 -356 286.5 -376Q320 -396 340 -432.5Q360 -469 360 -516Q360 -563 340 -599.5Q320 -636 286.5 -656.5Q253 -677 211 -677Q169 -677 135.5 -656.5Q102 -636 82.5 -599.5Q63 -563 63 -516Q63 -469 82.5 -432.5Q102 -396 135.5 -376Q169 -356 211 -356ZM139 -417V-620H215Q247 -620 270 -605.5Q293 -591 293 -555Q293 -537 283 -521.5Q273 -506 256 -500L302 -417H256L221 -487H180V-417ZM180 -519H206Q249 -519 249 -553Q249 -568 240 -577Q231 -586 208 -586H180Z"/>
</svg>
```
but the correct `viewBox` should be,

    "23 -713 377 394"

since it makes the SVG's boundaries be snug to the glyph.

---
If made a set of 11 test fonts that you can find inside this ZIP file: [fixtures.zip](https://github.com/adobe-type-tools/opentype-svg/files/6543689/fixtures.zip). Please add them to this PR, placing them inside a folder named **fixtures** which should be placed inside the existing **tests** folder.

Also add this snippet to **conftest.py**,
```python
@pytest.fixture
def fixtures_dir():
    yield os.path.join(os.path.dirname(__file__), 'fixtures')
```
and update the `test_adjust_to_viewbox` method to this,
```python
import os
import pytest


@pytest.mark.parametrize('file_name, expected', [
    ("1.ttf", "309 -1108 360 638"),
    ("2.ttf", "-904 -1110 416 650"),
    ("3.ttf", "-824 310 420 662"),
    ("4.ttf", "197 342 452 638"),
    ("12.ttf", "-904 -1110 1573 650"),
    ("13.ttf", "-824 -1108 1493 2080"),
    ("14.ttf", "197 -1108 472 2088"),
    ("23.ttf", "-904 -1110 500 2082"),
    ("24.ttf", "-904 -1110 1553 2090"),
    ("34.ttf", "-824 310 1473 670"),
    ("1234.ttf", "-904 -1110 1573 2090"),
])
def test_adjust_to_viewbox(file_name, expected, fixtures_dir):
    font_path = os.path.join(fixtures_dir, file_name)
    viewbox = viewbox_settings(font_path, True)
    assert viewbox == expected
```

I haven't had time to determine what changes are needed to produce the `width` and `height` values above, but I can assure you that they're correct.

**Comment by davidgodzsak** [comment]
But wouldn't I need the `hhea` table to get the `ascend`, `descend`, `linegap` to calculate the height and `xMaxExtent` or similar to get the `width` liek I did in the first iteration?

**Comment by miguelsousa** [comment]
> the tables in ttf tell you information about the whole font and not the glyphs themselves

Yes, the x/yMin/Max values in the`head` table define the combined bounding box of all glyphs in the font, not the bounding box of each individual glyph.

> With this, the borders won't snuggle the glyph

Right, in a common font they likely won't because usually not all glyphs will have the same bounding box.
But in the case of the test fonts, each of them has a single glyph, so the font's bbox will certainly match the glyph's. I did this to make it easy to validate the viewBox calculations.

> If the desired effect is to snuggle the fonts then that has to be dynamically calculated for every glyph.

For me, the goal of the new `-av` option is to make the tool output the SVGs with a viewBox that encloses all of the glyphs in the font. The viewBox value will be the same for all SVGs, and all the glyphs will be fully visible (as opposed to the default behavior which makes a "view window" of 1000x1000 edge-aligned at 0,0).

**Comment by davidgodzsak** [comment]
@miguelsousa I think I got it this time, thanks for all the help!

**Review by miguelsousa — APPROVED** [review]
👍 Thanks for contributing this functionality.
